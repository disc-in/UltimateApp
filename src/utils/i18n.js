import { I18n } from 'i18n-js';
import * as Localization from 'expo-localization';

import en from './locales/en';
import fr from './locales/fr';

const i18n = new I18n({
  en,
  fr,
});

i18n.defaultLocale = 'en';
i18n.locale = Localization.locale;
i18n.fallbacks = true;
i18n.enableFallback = true;

export default i18n;
