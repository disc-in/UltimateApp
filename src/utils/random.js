import * as Crypto from 'expo-crypto';

export const generateRandomHex = (size = 10) =>
  [...Array(size)].map(() => Math.floor(Math.random() * 16).toString(16)).join('');

export const generateUuid = () => Crypto.randomUUID();
