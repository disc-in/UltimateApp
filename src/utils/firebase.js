import { Platform, InteractionManager } from 'react-native';
import Constants from 'expo-constants';
import { initializeApp, setLogLevel } from 'firebase/app';
import { getDatabase, ref, set, child, get } from 'firebase/database';

import { generateRandomHex } from './random';

// Work around issue `Setting a timer for long time`
// see: https://github.com/firebase/firebase-js-sdk/issues/97
const _setTimeout = global.setTimeout;
const _clearTimeout = global.clearTimeout;
const MAX_TIMER_DURATION_MS = 60 * 1000;
if (Platform.OS === 'android') {
  const timerFix = {};
  const runTask = (id, fn, ttl, args) => {
    const waitingTime = ttl - Date.now();
    if (waitingTime <= 1) {
      InteractionManager.runAfterInteractions(() => {
        if (!timerFix[id]) {
          return;
        }
        delete timerFix[id];
        fn(...args);
      });
      return;
    }

    const afterTime = Math.min(waitingTime, MAX_TIMER_DURATION_MS);
    timerFix[id] = _setTimeout(() => runTask(id, fn, ttl, args), afterTime);
  };

  global.setTimeout = (fn, time, ...args) => {
    if (MAX_TIMER_DURATION_MS < time) {
      const ttl = Date.now() + time;
      const id = '_lt_' + Object.keys(timerFix).length;
      runTask(id, fn, ttl, args);
      return id;
    }
    return _setTimeout(fn, time, ...args);
  };

  global.clearTimeout = (id) => {
    if (typeof id === 'string' && id.startsWith('_lt_')) {
      _clearTimeout(timerFix[id]);
      delete timerFix[id];
      return;
    }
    _clearTimeout(id);
  };
}
// End of workaround

const firebaseConfig = {
  databaseURL: Constants.expoConfig.extra.firebaseDatabaseUrl,
};

const app = initializeApp(firebaseConfig);
setLogLevel('silent');

const reference = (namespace, identifier) => {
  return ref(getDatabase(app), `${namespace}/${identifier}`);
};

export const upload = async (namespace, record) => {
  const withoutUndefineds = JSON.parse(JSON.stringify(record));
  const identifier = generateRandomHex();
  await set(reference(namespace, identifier), withoutUndefineds);
  return identifier;
};

export const download = (namespace, identifier) => {
  const dbRef = ref(getDatabase(app));
  return get(child(dbRef, `${namespace}/${identifier}`))
    .then((snapshot) => {
      if (snapshot.exists()) {
        return snapshot.val();
      } else {
        return null;
      }
    })
    .catch((error) => {
      console.error(error);
    });
};
