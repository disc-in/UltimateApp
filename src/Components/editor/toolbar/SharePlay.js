import React, { useState, useRef } from 'react';
import { Platform, TouchableOpacity, Share } from 'react-native';
import { Ionicons } from '@expo/vector-icons';

import theme from '../../../styles/theme.style';
import I18n from '../../../utils/i18n';
import { showError } from '../../../utils/flashMessage';
import { upload } from '../../../utils/firebase';

const SharePlay = ({ currentPlay }) => {
  const share = async () => {
    try {
      const uuid = await upload('customPlays', currentPlay);
      await Share.share({
        title: I18n.t('editor.sharePlay.shareTitle', { title: currentPlay.title }),
        message: I18n.t('editor.sharePlay.shareMessage', { uuid }),
      });
    } catch (error) {
      showError(I18n.t('editor.sharePlay.shareError'));
    }
  };

  return (
    <TouchableOpacity onPress={share} testID="shareButton">
      <Ionicons
        name={Platform.select({
          ios: 'ios-share-outline',
          default: 'share-social-outline',
        })}
        color={theme.COLOR_PRIMARY_LIGHT}
        size={30}
      />
    </TouchableOpacity>
  );
};

export default SharePlay;
