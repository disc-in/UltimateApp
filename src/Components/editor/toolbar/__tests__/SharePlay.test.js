import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react-native';
import { Share } from 'react-native';

import animationSquare from '../../../../Fixtures/Animation/AnimationSquare';
import * as firebase from '../../../../utils/firebase';

import SharePlay from '../SharePlay';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.mock('react-native-get-random-values', () => ({
  getRandomBase64: jest.fn(),
}));

describe('<SharePlay />', () => {
  afterEach(() => jest.clearAllMocks());

  const currentPlay = { animation: animationSquare, title: 'Square', uuid: '123' };

  it('renders correctly', async () => {
    const { toJSON } = await waitFor(() => render(<SharePlay currentPlay={currentPlay} />));
    expect(toJSON()).toMatchSnapshot();
  });

  it('triggers share action', async () => {
    jest.spyOn(firebase, 'upload').mockImplementation(() => 'uuid');

    const share = jest.fn();
    Share.share = () => new Promise((resolve, reject) => share());

    const { getByTestId } = await waitFor(() => render(<SharePlay currentPlay={currentPlay} />));
    fireEvent.press(getByTestId('shareButton'));

    await expect(firebase.upload).toHaveBeenCalledWith('customPlays', currentPlay);
    expect(share).toHaveBeenCalled();
  });
});
