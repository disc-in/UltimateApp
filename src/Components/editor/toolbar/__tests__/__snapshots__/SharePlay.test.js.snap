// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SharePlay /> renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="shareButton"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 30,
        },
        undefined,
        Object {
          "fontFamily": "ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    ï”˜
  </Text>
</View>
`;
